lib/balanced-pool.js

```diff
@@ -41,6 +65,16 @@ class BalancedPool extends Dispatcher {
 
     pool.on('disconnect', (...args) => {
       this.emit('disconnect', ...args)
+
+      const err = args[2]
+      if (err) {
+        // decrease the weight of the pool.
+        pool[kWeight] = Math.max(1, pool[kWeight] - MAX_WEIGHT * 0.01)
+      } else {
+        // Set the weight of a connection to the maximum value.
+        pool[kWeight] = Math.round(MAX_WEIGHT / this[kPools].length)
```

Comment:
I feel this might be better done on successful connection.

Reply:
> I feel this might be better done on successful connection.

What if we have a server that drops the connections as soon as we get a connection? 

This is what i'm trying to cover. I'm also simulating that in the tests.


but, I also still have to cover this case: 

* Client is unable to connect to the upstream

Reply:
> > What if the upstream drops the connections as soon as the client connects?

> This is why I preferred gluing the logic on a disconnect event.

I'm ok in decreasing the weight if a connection is closed with an error, however I'm not in resetting the weight if a connection is disconnected successfully. It should either be reset on successful connection or at the first successful request?